File: scripts/monolith_watcher.py
#!/usr/bin/env python3
"""
M-One Anti-Monolith Watcher (scripts/monolith_watcher.py)
Monitora arquivos Python, HTML e JS do projeto M-One contra crescimento descontrolado.
Emite alertas coloridos no terminal e notificações nativas no macOS via osascript.

Uso:
  python3 scripts/monolith_watcher.py          # Executa varredura e fica observando alterações
  python3 scripts/monolith_watcher.py --scan   # Executa varredura única (auditoria) e sai
"""

import argparse
import os
import subprocess
import sys
import time
from pathlib import Path

LINE_LIMIT = 500
BASE_DIR = Path(__file__).resolve().parent.parent

# Diretórios e extensões monitoradas
WATCH_PATTERNS = [
    ("app.py",),
    ("database.py",),
    ("routes", "*.py"),
    ("templates", "*.html"),
    ("static/js", "*.js"),
]


def notify_macos(title: str, message: str):
    """Envia notificação nativa para o macOS com som de alerta."""
    try:
        script = f'display notification "{message}" with title "{title}" sound name "Glass"'
        subprocess.run(["osascript", "-e", script], check=False, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
    except Exception:
        pass


def count_lines(filepath: Path) -> int:
    try:
        with open(filepath, "r", encoding="utf-8", errors="ignore") as f:
            return sum(1 for _ in f)
    except Exception:
        return 0


def collect_target_files() -> list[Path]:
    files = []
    # Arquivos raiz
    for root_file in ["app.py", "database.py"]:
        p = BASE_DIR / root_file
        if p.is_file():
            files.append(p)

    # Diretórios
    for folder, pattern in [("routes", "*.py"), ("templates", "*.html"), ("static/js", "*.js")]:
        d = BASE_DIR / folder
        if d.is_dir():
            files.extend(d.glob(pattern))

    return sorted(set(files))


def check_file(filepath: Path, silent_if_ok: bool = True) -> bool:
    lines = count_lines(filepath)
    rel_path = filepath.relative_to(BASE_DIR)

    if lines > LINE_LIMIT:
        print(f"\n\033[41m\033[37m ⚠️  ALERTA DE MONÓLITO M-ONE ⚠️  \033[0m")
        print(f"\033[31mO arquivo \033[1m{rel_path}\033[0m\033[31m ultrapassou o limite estabelecido!\033[0m")
        print(f"Linhas atuais: \033[1m{lines}\033[0m / Limite: {LINE_LIMIT}")
        print(f"\033[33mRecomendação: Ativar skill de Refatoração Proativa para modularizar em Blueprints.\033[0m\n")

        notify_macos(
            "⚠️ M-One: Alerta de Monólito!",
            f"O arquivo {filepath.name} atingiu {lines} linhas. Considere modularizar.",
        )
        return False
    else:
        if not silent_if_ok:
            print(f"  ✓ {rel_path}: {lines} linhas (dentro do limite)")
        return True


def run_audit():
    print(f"\033[36m[Anti-Monolith Watcher]\033[0m Varrendo arquivos do projeto (limite: {LINE_LIMIT} linhas)...")
    files = collect_target_files()
    violators = 0

    for f in files:
        if not check_file(f, silent_if_ok=False):
            violators += 1

    print("\n" + "=" * 50)
    if violators == 0:
        print("\033[32m✅ Nenhum monólito crítico encontrado!\033[0m")
    else:
        print(f"\033[33m⚠️  Total de arquivos acima de {LINE_LIMIT} linhas: {violators}\033[0m")
    print("=" * 50 + "\n")


def watch_loop():
    print(f"\033[36m[Anti-Monolith Watcher]\033[0m Observando alterações em tempo real (Ctrl+C para sair)...")
    mtimes: dict[Path, float] = {}

    for f in collect_target_files():
        try:
            mtimes[f] = f.stat().st_mtime
        except OSError:
            pass

    while True:
        try:
            time.sleep(2)
            current_files = collect_target_files()
            for f in current_files:
                try:
                    current_mtime = f.stat().st_mtime
                    last_mtime = mtimes.get(f)
                    if last_mtime is None or current_mtime > last_mtime:
                        mtimes[f] = current_mtime
                        check_file(f, silent_if_ok=True)
                except OSError:
                    pass
        except KeyboardInterrupt:
            print("\n[Anti-Monolith Watcher] Encerrado.")
            break


def main():
    parser = argparse.ArgumentParser(description="M-One Anti-Monolith Watcher")
    parser.add_argument("--scan", action="store_true", help="Executa varredura única e finaliza")
    args = parser.parse_args()

    run_audit()
    if not args.scan:
        watch_loop()


if __name__ == "__main__":
    main()

File: scripts/monolith-watcher.js
#!/usr/bin/env node
/**
 * M-One Anti-Monolith Watcher (scripts/monolith-watcher.js)
 * Monitora em tempo real arquivos do projeto M-One.
 * Emite alertas sonoros e notificações nativas do macOS/Windows caso arquivos ultrapassem o limite de linhas.
 */

const fs = require('fs');
const path = require('path');

let chokidar, notifier;
try {
  chokidar = require('chokidar');
  notifier = require('node-notifier');
} catch (e) {
  console.log('\x1b[33m[Aviso]\x1b[0m Módulos "chokidar" ou "node-notifier" não encontrados localmente.');
  console.log('Execute: npm install --no-save chokidar node-notifier');
  console.log('Ou utilize o watcher nativo em Python: python3 scripts/monolith_watcher.py\n');
}

const LINE_LIMIT = 500;
const BASE_DIR = path.resolve(__dirname, '..');

const WATCH_PATHS = [
  path.join(BASE_DIR, 'app.py'),
  path.join(BASE_DIR, 'database.py'),
  path.join(BASE_DIR, 'routes/**/*.py'),
  path.join(BASE_DIR, 'templates/**/*.html'),
  path.join(BASE_DIR, 'static/js/**/*.js')
];

console.log(`\x1b[36m[Anti-Monolith Watcher - M-One]\x1b[0m Iniciado! Observando arquivos (> ${LINE_LIMIT} linhas)...`);

function checkFile(filePath) {
  try {
    if (!fs.existsSync(filePath)) return;
    const stat = fs.statSync(filePath);
    if (!stat.isFile()) return;

    const content = fs.readFileSync(filePath, 'utf8');
    const lineCount = content.split('\n').length;

    if (lineCount > LINE_LIMIT) {
      const fileName = path.basename(filePath);
      const relPath = path.relative(BASE_DIR, filePath);

      // Alerta no terminal
      console.log(`\n\x1b[41m\x1b[37m ⚠️  ALERTA DE MONÓLITO M-ONE ⚠️  \x1b[0m`);
      console.log(`\x1b[31mO arquivo \x1b[1m${relPath}\x1b[0m\x1b[31m ultrapassou o limite estabelecido!\x1b[0m`);
      console.log(`Linhas atuais: \x1b[1m${lineCount}\x1b[0m / Limite: ${LINE_LIMIT}`);
      console.log(`\x1b[33mRecomendação: Ativar skill de Refatoração Proativa para modularizar em rotas/Blueprints.\x1b[0m\n`);

      // Notificação nativa no OS (Mac/Windows)
      if (notifier) {
        notifier.notify({
          title: '⚠️ M-One: Alerta de Monólito!',
          message: `O arquivo ${fileName} atingiu ${lineCount} linhas. Considere modularizar em Blueprints.`,
          sound: true,
          wait: false
        });
      }
    }
  } catch (error) {
    console.error(`\x1b[31m[Anti-Monolith Watcher] Erro ao analisar ${filePath}:\x1b[0m`, error.message);
  }
}

if (chokidar) {
  const watcher = chokidar.watch(WATCH_PATHS, {
    ignored: /(^|[\/\\])\../,
    persistent: true,
    ignoreInitial: false
  });

  watcher
    .on('add', filePath => checkFile(filePath))
    .on('change', filePath => checkFile(filePath));
} else {
  // Fallback sem chokidar: varredura direta uma vez
  console.log('Executando varredura estática de arquivos...');
  const filesToCheck = [
    path.join(BASE_DIR, 'app.py'),
    path.join(BASE_DIR, 'database.py')
  ];
  filesToCheck.forEach(checkFile);
}
